Added functions to ArrayComprehension

kangzhiq opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed
- Added .tolist() and tests
- Added .tomatrix() and tests
- Refactorised _expand_array()


#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* tensor
  * Added functions to ArrayComprehension.
<!-- END RELEASE NOTES -->

Commit:
Added functions to ArrayComprehension

- Added .tolist() and tests
- Added .tomatrix() and tests
- Refactorised _expand_array()

czgdp1807 commented on sympy/tensor/array/array_comprehension.py:
```diff
@@ -272,4 +273,59 @@ def _array_subs(arr, var, val):
                 else:
                     list_gen.append(_array_subs(list_expr, var, val))
-        return ImmutableDenseNDimArray(list_gen)
+        return list_gen
```
Why `ImmutableDenseNDimArray` is shifted to `doit()`? Something related to performance?

czgdp1807 replied:
Ah! got it, it is being used in `tomatrix`, etc. So returning a generalized expression for type casting. No issues. Resolve this conversation.

czgdp1807 commented on sympy/tensor/array/array_comprehension.py:
```diff
@@ -272,4 +273,59 @@ def _array_subs(arr, var, val):
+
+        if not self.is_numeric:
+            raise ValueError("A symbolic array cannot be expanded to a matrix")
+        if self.rank() != 2:
```
I think there was something like, `self._rank`, can it be used here directly? As far as I remember `rank` also calls `self._rank`. A short explanation would help if the said cannot be done. Thanks.

kangzhiq replied:
Yes, There is nearly no difference between `self._rank` and `self.rank()`. BUt it would be better to keep the same coding convention. So that is a mistake. Thank for pointing it out!

Upabjojr replied:
OK

czgdp1807 commented:
I think `Added functions to ArrayComprehension` should be replaced by `Added methods to ArrayComprehension`. I don't know if I am wrong conceptually.

kangzhiq commented:
> be replaced by `Added methods to ArrayComprehension`

I am sorry but I don't get your point. Do you mean this would lead to a confusion with the `function` in `ArrayComprehension`? But I would say `functions` is still comprehensible as a title of PR. So I would prefer keeping it as it has been done.

kangzhiq commented on sympy/tensor/array/array_comprehension.py:
```diff
@@ -272,4 +273,59 @@ def _array_subs(arr, var, val):
+        >>> a.tolist()
+        [[11, 12, 13], [21, 22, 23], [31, 32, 33], [41, 42, 43]]
+        """
+        if self.is_numeric:
```
@Upabjojr Regarding this `.is_numeric` property(dicussion raised in #16929), since I am reusing it here, it would be an occasion to modify it.
I have searched over the SymPy API and it is true that this is the only case using `is_numeric`. So maybe it is better to follow the standard. :-)

Upabjojr replied:
What about "is_shape_numeric"?

kangzhiq replied:
@Upabjojr That would be a little bit more informative than `is_numeric`. But it is not used in SymPy either. Anyway, I will go with `is_shape_numeric` if you are ok with it. :-)

Upabjojr commented on sympy/tensor/array/tests/test_array_comprehension.py:
```diff
@@ -40,3 +46,5 @@ def test_array_comprehension():
     raises(ValueError, lambda: len(ArrayComprehension(i*j, (i, 1, 3), (j, 2, j+4))))
     raises(TypeError, lambda: ArrayComprehension(i*j, (i, 0, i + 1.5), (j, 0, 2)))
+    raises(ValueError, lambda: b.tolist())
+    raises(ValueError, lambda: b.tomatrix())
```
also check this exception on numeric shape and rank different than 2.

Commit:
Renamed is_numeric and added tests

- Renamed is_numeric to is_shape_numeric
- Modified and added tests accordingly
